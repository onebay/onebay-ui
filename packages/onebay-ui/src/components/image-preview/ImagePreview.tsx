import classNames from 'classnames'
import { defineComponent, PropType, ref, computed, watch } from 'vue'
import Swiper from '../swiper'
import { getLen } from '../../utils'

const threshold = 10

const ImagePreview = defineComponent({
  name: 'ImagePreview',
  components: {
    Swiper
  },
  props: {
    className: {
      type: String,
      default: ''
    },
    images: {
      type: Array as PropType<string[]>,
      default: () => []
    },
    show: {
      type: Boolean,
      default: false
    },
    startPosition: {
      type: Number,
      default: 1
    },
    paginationVisible: {
      type: Boolean,
      default: true
    },
    maxScale: {
      type: Number,
      default: 3
    },
    minScale: {
      type: Number,
      default: 1
    }
  },
  emits: ['close', 'open'],
  setup(props, { emit }) {
    const touchPoint = []
    let movePoint = []
    const scale = ref(1)
    const visible = ref(props.show)
    watch(
      () => props.show,
      (val) => {
        /* istanbul ignore else */
        if (val !== visible.value) {
          visible.value = val
        }
      }
    )
    watch(
      () => visible.value,
      (val) => {
        if (val) {
          emit('open')
        } else {
          emit('close')
        }
      }
    )
    const toggleVisible = () => {
      visible.value = !visible.value
    }
    const handleTouchstart = (event: TouchEvent) => {
      const touchList = [].slice.call(event.targetTouches)
      /* istanbul ignore else */
      if (touchList.length > 1) {
        touchList.forEach((item: Touch, index: number) => {
          const point = {
            index: item.identifier,
            x: item.clientX,
            y: item.clientY
          }
          touchPoint[index] = point
        })
        movePoint = JSON.parse(JSON.stringify(touchPoint))
      }
    }
    const handleTouchmove = (event: TouchEvent) => {
      const { minScale, maxScale } = props
      const touchList = [].slice.call(event.changedTouches)
      touchList.forEach((item: Touch, index: number) => {
        const point = {
          index: item.identifier,
          x: item.clientX,
          y: item.clientY
        }
        const matchPoint = movePoint.find((item) => item.index === point.index)
        if (matchPoint) {
          matchPoint.x = point.x
          matchPoint.y = point.y
        } else {
          movePoint[index] = point
        }
      })
      /* istanbul ignore else */
      if (movePoint.length > 1) {
        const start = getLen(touchPoint)
        const end = getLen(movePoint)
        /* istanbul ignore else */
        if (threshold < Math.abs(start - end)) {
          const temp = end / start
          if (temp > maxScale) {
            scale.value = maxScale
          } else if (temp < minScale) {
            scale.value = minScale
          } else {
            scale.value = temp
          }
        }
      }
    }
    const handleTouchend = (event: TouchEvent) => {
      const idx1 = touchPoint.findIndex((item) => item.index === event.changedTouches[0].identifier)
      touchPoint.splice(idx1, 1)
      const idx2 = movePoint.findIndex((item) => item.index === event.changedTouches[0].identifier)
      movePoint.splice(idx2, 1)
    }
    const scaleStyle = computed(() => {
      return {
        transform: `scale(${scale.value}, ${scale.value})`
      }
    })
    const handleDbClick = () => {
      if (scale.value > 1) {
        scale.value = 1
      } else {
        scale.value = 1.5
      }
    }

    return {
      handleTouchstart,
      handleTouchmove,
      handleTouchend,
      scaleStyle,
      handleDbClick,
      visible,
      toggleVisible,
      scale
    }
  },
  render() {
    const { className, images, startPosition, paginationVisible } = this.$props
    const {
      handleTouchstart,
      handleTouchmove,
      handleTouchend,
      scaleStyle,
      handleDbClick,
      visible,
      toggleVisible
    } = this
    return (
      <div
        class={classNames('ob-image-preview', className, { 'ob-image-preview-active': visible })}
        onTouchstart={handleTouchstart}
        onTouchmove={handleTouchmove}
        onTouchend={handleTouchend}
        onDblclick={handleDbClick}>
        <div class="ob-image-preview-mask" onClick={toggleVisible}></div>
        {visible && (
          <Swiper
            swiperData={images}
            class="ob-image-preview-content"
            style={scaleStyle}
            initPage={startPosition}
            paginationVisible={paginationVisible}>
            {images.map((item: string, index: number) => (
              <div key={index} class="ob-swiper-slide">
                <img src={item} style="max-width: 100%" />
              </div>
            ))}
          </Swiper>
        )}
      </div>
    )
  }
})

export default ImagePreview
