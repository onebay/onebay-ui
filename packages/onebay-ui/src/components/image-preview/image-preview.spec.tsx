import { mount } from '@vue/test-utils'
import ImagePreview from './index'

const factory = (values = {}, slots = { default: [] }) => {
  return mount(ImagePreview, {
    slots,
    props: { ...values },
  })
}

const files = [
  'https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg',
  'https://zos.alipayobjects.com/rmsportal/hqQWgTXdrlmVVYi.jpeg',
  'https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg',
  'https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg',
  'https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg',
]
describe('ImagePreview Snap', () => {

  it('render initial ImagePreview', () => {
    const wrapper = factory()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render ImagePreview props images', () => {
    const wrapper = factory({ images: files })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render ImagePreview props show', () => {
    const wrapper = factory({ images: files, show: true })
    expect(wrapper.element).toMatchSnapshot()
  })
})

describe('ImagePreview Event', () => {
  it('ImagePreview click mask', async () => {
    const wrapper = factory({ images: files, show: true })
    expect(wrapper.vm.visible).toBe(true)
    await wrapper.find('.ob-image-preview-mask').trigger('click')
    expect(wrapper.vm.visible).toBe(false)
    expect(wrapper.emitted()).toHaveProperty('close')
  })

  it('ImagePreview emit open event', async () => {
    const wrapper = factory({ images: files })
    await wrapper.setProps({ show: true })
    expect(wrapper.emitted()).toHaveProperty('open')
  })

  it('ImagePreview dblclick event', () => {
    const wrapper = factory({ images: files, show: true })
    wrapper.trigger('dblclick')
    expect(wrapper.vm.scale).toEqual(1.5)
    wrapper.trigger('dblclick')
    expect(wrapper.vm.scale).toEqual(1)
  })

  it('ImagePreview test touch event', () => {
    const wrapper = factory({ images: files, show: true })
    wrapper.trigger('touchstart', {
      targetTouches: [
        { identifier: 0, clientX: 50, clientY: 50 },
        { identifier: 1, clientX: 100, clientY: 100 },
      ]
    })
    // test minScale
    wrapper.trigger('touchmove', {
      changedTouches: [
        { identifier: 0, clientX: 40, clientY: 40 },
        { identifier: 1, clientX: 60, clientY: 60 }
      ]
    })
    wrapper.trigger('touchmove', {
      changedTouches: [
        { identifier: 0, clientX: 40, clientY: 40 },
        { identifier: 1, clientX: 120, clientY: 120 },
        { identifier: 2, clientX: 120, clientY: 120 }
      ]
    })
    // test maxScale
    wrapper.trigger('touchmove', {
      changedTouches: [
        { identifier: 0, clientX: 40, clientY: 40 },
        { identifier: 1, clientX: 320, clientY: 320 }
      ]
    })
    wrapper.trigger('touchend', {
      changedTouches: [
        { identifier: 0, clientX: 40, clientY: 40 },
        { identifier: 1, clientX: 120, clientY: 120 },
      ]
    })
  })
})