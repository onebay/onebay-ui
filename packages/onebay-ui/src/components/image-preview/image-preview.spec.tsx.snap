// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImagePreview Snap render ImagePreview props images 1`] = `
<div
  class="ob-image-preview"
>
  <div
    class="ob-image-preview-mask"
  />
  <!---->
</div>
`;

exports[`ImagePreview Snap render ImagePreview props show 1`] = `
<div
  class="ob-image-preview ob-image-preview-active"
>
  <div
    class="ob-image-preview-mask"
  />
  <div
    class="ob-swiper horizontal ob-image-preview-content"
    style="transform: scale(1, 1);"
  >
    <div
      class="ob-swiper-wrap"
      style="transform: translate3d(0px,0px,0); transition-duration: 0ms; transition-timing-function: ease;"
    >
      
      
      <div
        class="ob-swiper-slide"
      >
        <img
          src="https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg"
          style="max-width: 100%;"
        />
      </div>
      <div
        class="ob-swiper-slide"
      >
        <img
          src="https://zos.alipayobjects.com/rmsportal/hqQWgTXdrlmVVYi.jpeg"
          style="max-width: 100%;"
        />
      </div>
      <div
        class="ob-swiper-slide"
      >
        <img
          src="https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg"
          style="max-width: 100%;"
        />
      </div>
      <div
        class="ob-swiper-slide"
      >
        <img
          src="https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg"
          style="max-width: 100%;"
        />
      </div>
      <div
        class="ob-swiper-slide"
      >
        <img
          src="https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg"
          style="max-width: 100%;"
        />
      </div>
      
      
    </div>
    <div
      class="ob-swiper-pagination"
    >
      
      
    </div>
  </div>
</div>
`;

exports[`ImagePreview Snap render initial ImagePreview 1`] = `
<div
  class="ob-image-preview"
>
  <div
    class="ob-image-preview-mask"
  />
  <!---->
</div>
`;
